plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/async/PermitBackedRetryableFutureUtils.java

```diff
@@ -0,0 +1,344 @@
+        if (!retryBackoffDelayIterator.hasNext()) {
+            return CompletableFuture.failedFuture(ex);
+        }
+
+        return flatten(
+            flatten(
+                CompletableFuture.supplyAsync(
+                    futureSupplier,
+                    new DelayedExecutor(
+                        retryableException.retryBackoffDelayIterator.next().millis(),
+                        TimeUnit.MILLISECONDS,
+                        remoteTransferRetryPool,
+                        scheduler
+                    )
+                )
+            ).thenApply(CompletableFuture::completedFuture).exceptionally(t -> {
+                if (t instanceof RetryableException) {
+                    ex.addSuppressed(t);
+                    return retry(ex, futureSupplier, retryBackoffDelayIterator);
+                } else {
+                    ex.addSuppressed(t);
+                    return CompletableFuture.failedFuture(ex);
+                }
+            })
+        );
+    }
+
+    // Package access for testing
+    Semaphore acquirePermit(WritePriority writePriority, boolean isLowPriorityPermitsConsumable) throws InterruptedException {
+        // Try acquiring low priority permit or high priority permit immediately if available.
+        // Otherwise, we wait for low priority permit.
+        if (Objects.requireNonNull(writePriority) == WritePriority.LOW) {
+            if (lowPrioritySemaphore.tryAcquire()) {
+                return lowPrioritySemaphore;
+            } else if (highPrioritySemaphore.availablePermits() > 0.4 * highPriorityPermits && highPrioritySemaphore.tryAcquire()) {
+                return highPrioritySemaphore;
+            } else if (lowPrioritySemaphore.tryAcquire(5, TimeUnit.MINUTES)) {
+                return lowPrioritySemaphore;
+            }
+            return null;
+        }
+
+        // Try acquiring high priority permit or low priority permit immediately if available.
+        // Otherwise, we wait for high priority permit.
+        if (highPrioritySemaphore.tryAcquire()) {
+            return highPrioritySemaphore;
+        } else if (isLowPriorityPermitsConsumable && lowPrioritySemaphore.tryAcquire()) {
+            return lowPrioritySemaphore;
+        } else if (highPrioritySemaphore.tryAcquire(5, TimeUnit.MINUTES)) {
+            return highPrioritySemaphore;
+        }
+        return null;
+    }
+
+    private Supplier<CompletableFuture<FutureType>> createPermitBackedFutureSupplier(
+        Iterator<TimeValue> retryBackoffDelayIterator,
+        boolean lowPriorityPermitsConsumable,
+        Supplier<CompletableFuture<FutureType>> futureSupplier,
+        WritePriority writePriority
+    ) {
+        return () -> {
+            Semaphore semaphore;
+            try {
+                semaphore = acquirePermit(writePriority, lowPriorityPermitsConsumable);
+                if (semaphore == null) {
+                    throw new RetryableException(retryBackoffDelayIterator);
```

Comment:
Why are we preferring semaphore based approach instead of a queue based producer-consumer approach ? Just looking for Pros-Cons analysis here . The exception handling and retries are kind of misleading here . Instead we can model it around `FixedThreadPool` kind of approach, where we reject the excess work pre emptively .

Reply:
Additional queue can only help in handling scenario where huge number of batch of files are requested simultaneously but semaphore would be required in any case to ensure controlled transfers. Using a thread pool would be counter productive because firstly we don't want to make it async and instead want to build a backpressure on main thread and secondly, there would be additional unnecessary overhead of context switch.

How are retries and exception handling misleading? They do work. Take a look at the test cases.

Reply:
I am trying to understand what are the current gaps with S3 standard SDK when it comes to limiting concurrency and are their opportunities to contribute back. For instance we are re-introducing permit acquisition timeout which seems analogous to connection acquisition timeouts and to achieve optimal results both have to be in-sync for a vast combination of instance types.
Permits would have made sense if we were uploading a single file iff enough permits for all parts were available to ensure we prioritise upload of one file in a bid to minimise part failures across multiple files and retries needed across multiple files which would have caused responses from other parts to be held up in memory.
We need a scheme which queues up files to be uploaded but don't break them into parts unless enough permits are available so that the memory overhead of a file yet to be uploaded is minimal.

Reply:
Take an example of a 5gb file which is available for upload. It would require 320 parts and hence 320 permits if we plan to break it only when permits are available. With current calculation that would require 320/5=64 processors for upload. With this approach we will be mostly performing single upload of a file and hence, won't be efficient. 

The problem we are trying to solve here is timeouts happening as a result of upload bursts. Problem is that as soon as an upload is submitted, S3 SDK puts into the connection queue and hence, a timer is started and in case of large queues, request times out even before it is attempted.

>which seems analogous to connection acquisition timeouts

With this approach we will have the opportunity to set connection and request timeouts to much lower reasonable numbers which today are set exceedingly high to provide opportunity for uploads to eventually succeed. This approach doesn't always work and we rely on manual retries for successful uploads. 

From memory consumption perspective, this approach makes sure that no upload begins (start consuming memory) till an existing upload is completed from start to end. This will keep memory consumption of uploads in check. We can extend this approach to downloads as well.

Two major benefits we get out this is back pressure on main thread in case of upload bursts and preventing unnecessary resource consumption due to uploads competing with each other for acquisition of a thread.

Queue would help if we have large number of batches of files submitted for upload at once. But for a single batch of files, we won't need it since backpressure on main thread would take care of slowing down upload. I tested it on a r7g.medium instance to recover split shards of a 180gb index with OSB of so workload running in parallel and didn't observe any IO error. I also carried out 2 perfs if there's any impact on performance but didn't observe any impact.

Reply:
Not sure if I was clear. The point I was making is we only start limited uploads of files based on permits, essentially if there are 100 permits around we start the download of one file with 320 parts(each part containing one permit) while placing other parts in a prioritised queue, ensuring file gets uploaded in a prioritised order to make sure replication delay of in-flight checkpoint is minimal.

I am not advocating for queue or permit, what I think is we might need both based on the requirements. Happy to discuss more offline